linode/manager
# upcoming: [DI-19328] - Added capability to save & retrieve CloudPulse user preferences

## Description 📝
Added a feature to save the users global filter selection to preferences & load it back when user revisit the page.


## Changes  🔄
Added UserPreference utility file to handle all the preferences related operations
Modified all the global filter components to by default get selected with preferences value, if any present.

## Target release date 🗓️
Please specify a release date to guarantee timely review of this PR. If exact date is not known, please approximate and update it as needed.

## Preview 📷
**Include a screenshot or screen recording of the change**

:bulb: Use `<video src="" />` tag when including recordings in table.

| Before  | After   |
| ------- | ------- |
| ![Screenshot 2024-07-04 at 3 45 50 PM](https://github.com/linode/manager/assets/165884194/cdc3ae02-ad24-49f3-808c-fc9ece56c8a1) | ![Screenshot 2024-07-04 at 3 46 09 PM](https://github.com/linode/manager/assets/165884194/b334de21-9438-4642-8016-f2c9bcaa5d84) |
| ![Screenshot 2024-07-04 at 3 46 29 PM](https://github.com/linode/manager/assets/165884194/8413619f-dc02-483b-9f19-848c76845ac3) | ![Screenshot 2024-07-04 at 3 46 44 PM](https://github.com/linode/manager/assets/165884194/0fabe0d9-0ebc-4e3e-bc0e-91103e0db7c0) |
| ![Screenshot 2024-07-02 at 3 46 27 PM](https://github.com/linode/manager/assets/165884194/a9bb91e7-b6eb-4a67-8c61-f9743db3bc2a) | ![Screenshot 2024-07-02 at 3 46 41 PM](https://github.com/linode/manager/assets/165884194/e78bcd7b-93a6-42dd-a97d-6739523445fe) |
| ![Screenshot 2024-07-02 at 3 48 02 PM](https://github.com/linode/manager/assets/165884194/780c1fde-be58-45a9-8e8a-916c08a0e4c9) | ![Screenshot 2024-07-02 at 3 48 09 PM](https://github.com/linode/manager/assets/165884194/20a4970f-feb8-454b-8423-becac979aec5) |

## How to test 🧪

Note: Since Dashboards api is not ready to use, therefore have to enable mock services for testing. But with mock services preferences will not be saved properly, that's why it can be verified from the network tab that preferences api call is going on initial loading of the CloudPulse page to retrieve the saved preferences as well as on changing of every global filter value to update it.

1. Select values from global filter drop-downs.
2. Refresh the page or open in new tab or new window
3. You will see that your previous selections are retrieved from the preferences ( no saved preferences will be retrieved in case of mock user).


## As an Author I have considered 🤔

*Check all that apply*
